### PR TITLE
BUG: view with fieldless dtype should raise if itemsize != 0

### LINKED ISSUE #14344: viewing an array with a fieldless structured dtype does not update the dtype:
As discovered in #14310 , viewing an array with a fieldless structured dtype does not update the dtype/itemsize:

```python
>>> a = np.zeros(2, dtype={'names':[], 'formats':[], 'offsets':[], 'itemsize':8})
>>> b = a.view(np.dtype([]))
>>> b
array([], shape=(2, 0),
      dtype={'names':[], 'formats':[], 'offsets':[], 'itemsize':8})
>>> b.dtype == np.dtype([])
False
```

### PR DESCRIPTION
**Pull request by ahaldane**
Fixes #14344

Minor bugfix caused by treating the "dtype([])" structured type as if it was the "V0" flexible type.

### CODE REVIEW

**Comment by eric-wieser** [comment]
I think this now warrants a few more tests, since this fixed a lot of bugs.
Here's another that I think is fixed
```
>>> np.dtype((np.dtype([]), 10))
dtype({'names':[], 'formats':[], 'offsets':[], 'itemsize':10})  # before
dtype(([], (10,)))  # after
```

**Comment by eric-wieser** [comment]
Probably worth a `..versionchanged` in `doc/source/reference/c-api/array.rst` too, where this macro is documented.

**Comment by eric-wieser** [comment]
Another fix, `np.fromiter((() for i in range(10)), [])` doesn't raise

File: numpy/core/src/multiarray/arrayobject.c [hunk 1]
```diff
@@ -1200,7 +1200,8 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
             }
         }
         if (res == NULL && !PyErr_Occurred()) {
-            PyErr_SetString(PyExc_ValueError, "No fields found.");
+            /* these dtypes had no fields */
+            return cmp_op == Py_EQ ? Py_True : Py_False;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/arrayobject.c]: This change isn't correct - it doesn't produce a correctly broadcast array. Let's leave this to a future PR - the rest looks great.

**Comment by ahaldane** [comment]
Thanks for great reviews, as usual.

I updated `PyDataType_ISUNSIZED`, and went through all the places it was used to check if this introduced any bugs. It did, which I fixed, and I added appropriate tests.

I have some hesitation on doing it this way, so consider this a WIP. First, in some of the places calling `ISUNSIZED` the code really did want to check just `elsize == 0`, fields or no fields. Also, with the same logic it seems like `PyDataType_ISFLEXIBLE` should also be False for structured 0-size types. However, when I make that change I get a bunch of segfaults, especially in the pickling code. .

File: numpy/core/src/multiarray/arrayobject.c (hunk 1 above)
Comment (ahaldane) [reply numpy/core/src/multiarray/arrayobject.c]: Fixed by falling through to the string compare case which does the right thing here.

File: numpy/core/src/multiarray/arrayobject.c [hunk 2]
```diff
@@ -1199,15 +1199,13 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
...
-            PyErr_SetString(PyExc_ValueError, "No fields found.");
-        }
         return res;
     }
     else {
         /*
-         * compare as a string. Assumes self and
-         * other have same descr->type
+         * compare as a string. Assumes self and other have same descr->type.
+         * Note that fieldless structured datatypes (itemsize 0) fall through
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/arrayobject.c]: What about fieldless types with padding?

File: numpy/core/src/multiarray/arrayobject.c [hunk 3]
```diff
@@ -1200,15 +1200,18 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
             }
         }
         if (res == NULL && !PyErr_Occurred()) {
-            PyErr_SetString(PyExc_ValueError, "No fields found.");
+            /* these dtypes had no fields */
+            res = PyArray_NewLikeArray(self, NPY_ANYORDER,
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/arrayobject.c]: This doesn't broadcast with `other`, unless the broadcasting has already happened here

Comment (ahaldane) [reply numpy/core/src/multiarray/arrayobject.c]: Ah, yeah. I think I've got it finally by using a multiiter.

File: numpy/core/src/multiarray/arrayobject.c [hunk 4]
```diff
@@ -1200,15 +1200,28 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
             }
         }
         if (res == NULL && !PyErr_Occurred()) {
-            PyErr_SetString(PyExc_ValueError, "No fields found.");
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/arrayobject.c]: Fixes gh-13438, I think

File: numpy/core/src/multiarray/methods.c [hunk 5]
```diff
@@ -1861,7 +1861,7 @@ array_reduce_ex(PyArrayObject *self, PyObject *args)
         PyDataType_FLAGCHK(descr, NPY_ITEM_HASOBJECT) ||
         (PyType_IsSubtype(((PyObject*)self)->ob_type, &PyArray_Type) &&
          ((PyObject*)self)->ob_type != &PyArray_Type) ||
-        PyDataType_ISUNSIZED(descr)) {
+        descr->elsize == 0) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/methods.c]: It's not clear to me why this condition is needed at all, although it admittedly looks harmless.

File: doc/source/reference/c-api/array.rst [hunk 6]
```diff
@@ -997,6 +997,9 @@ argument must be a :c:type:`PyObject *<PyObject>` that can be directly interpret
     called on flexible dtypes. Types that are attached to an array will always
     be sized, hence the array form of this macro not existing.
 
+    ..versionchanged:: 1.18.0
+    For structured datatypes with no fields this function now return False.
```
Comment (eric-wieser) [inline doc/source/reference/c-api/array.rst]: Does this line need to be indented?

typo: `returns`

**Review by eric-wieser — APPROVED** [review]
Looks pretty good - one unimportant question, and a doc typo. Getting the comparison fix in here should have enabled me to push some of my remaining fixes, which were untestable without it - thanks!

**Comment by eric-wieser** [comment]
I think an `upcoming_changes/14393.c_api.rst` note might be warranted about the macro behavior

**Comment by charris** [comment]
@ahaldane Ping.

**Comment by ahaldane** [comment]
About to push as you wrote that!

File: numpy/core/src/multiarray/methods.c (hunk 5 above)
Comment (ahaldane) [reply numpy/core/src/multiarray/methods.c]: not clear to me either. These lines previously discussed here: https://github.com/numpy/numpy/pull/12748/files#r249258808 and introduced in #12011, but not discussed there.

File: doc/source/reference/c-api/array.rst (hunk 6 above)
Comment (ahaldane) [reply doc/source/reference/c-api/array.rst]: Fixed typo. I indented it the same way as the `..versionadded` in this file so it should be ok.

**Comment by mattip** [comment]
Thanks @ahaldane and reviewers

**Pull request merged**